File: src/main/java/com/tkr/desk/DeskRegistry.java
package com.tkr.desk;

import com.tkr.ledger.AuditSpool;
import com.tkr.types.WireTypes;
import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

/**
 * Registry orchestrating desk module passes on batch frames.
 * Ported from legacy-cpp desk_registry.cc.
 */
public final class DeskRegistry {

    public static final class DeskRunContext {
        public int batchId;
        public int recordCount;
        public int totalQtyMilli;
        public boolean marginOk;
        public boolean complianceOk;
        public boolean settlementOk;
        public boolean allocationOk;
    }

    private final ProRataAllocator proRata = new ProRataAllocator();
    private final AllocationSolver solver = new AllocationSolver();
    private final ComplianceRuleEngine compliance = new ComplianceRuleEngine();
    private final MarginAggregator margin = new MarginAggregator();
    private final HaircutCalculator haircut = new HaircutCalculator();
    private final CorporateActionAdjuster corpAction = new CorporateActionAdjuster();
    private final RestrictionFilter restrictions = new RestrictionFilter();
    private final FeeAccrualEngine fees = new FeeAccrualEngine();
    private final PositionLedger positions = new PositionLedger();
    private final CounterpartyLimit counterpartyLimit = new CounterpartyLimit();
    private final SettlementCalendar calendar = new SettlementCalendar();
    private final LotSplitter lotSplitter = new LotSplitter();
    private final TaxLotMatcher taxMatcher = new TaxLotMatcher();
    private final BenchmarkTracker benchmark = new BenchmarkTracker();

    public Status runBatchDesks(BatchWireFrame frame, DeskRunContext ctx) {
        if (frame == null || ctx == null) return Status.BOUNDS_ERROR;
        ctx.batchId = frame.header.deskId;
        ctx.recordCount = frame.records.size();
        long totalQty = 0;
        for (WireBatchRecord rec : frame.records) totalQty += rec.qtyMilli;
        ctx.totalQtyMilli = (int) totalQty;

        Status st = runCompliancePass(frame);
        ctx.complianceOk = st == Status.OK;
        if ((frame.header.flags & WireTypes.BATCH_FLAG_COMPLIANCE_HOLD) != 0 && !ctx.complianceOk) return st;

        CounterpartyLimit.CounterpartyLimitResult cpResult = counterpartyLimit.evaluate(frame);
        if (cpResult.status == Status.COMPLIANCE_REJECT) {
            ctx.complianceOk = false;
            return Status.COMPLIANCE_REJECT;
        }

        SettlementCalendar.SettlementCalendarResult settleResult = calendar.computeBatch(frame);
        ctx.settlementOk = settleResult.status == Status.OK;
        if (settleResult.status != Status.OK) return settleResult.status;

        LotSplitter.LotSplitterResult split = lotSplitter.splitBatch(frame, LotSplitter.LotSplitPolicy.PRO_RATA_BY_WEIGHT);
        if (split.status != Status.OK) return split.status;

        ProRataAllocator.ProRataAllocatorResult alloc = proRata.allocate(frame);
        ctx.allocationOk = alloc.status == Status.OK;
        if (alloc.status != Status.OK) return alloc.status;

        int targetTotal = 0;
        for (WireBatchRecord rec : frame.records) targetTotal += rec.qtyMilli;
        AllocationSolver.AllocationSolverResult solved = solver.solveWithBounds(alloc.slices, null, targetTotal);
        if (solved.status != Status.OK) return solved.status;

        Status ledgerSt = positions.applyBatch(frame, solved.finalSlices);
        if (ledgerSt != Status.OK) return ledgerSt;

        FeeAccrualEngine.FeeAccrualEngineResult feeRes = fees.accrueBatch(frame);
        if (feeRes.status != Status.OK) return feeRes.status;

        for (WireBatchRecord rec : frame.records) {
            TaxLotMatcher.TaxLot lot = new TaxLotMatcher.TaxLot();
            lot.lotId = rec.recordId;
            lot.accountId = rec.accountId;
            lot.symbolId = rec.symbolId;
            lot.qtyMilli = rec.qtyMilli;
            lot.costBasisCents = (long) rec.qtyMilli * rec.priceTick / 1000L;
            lot.acquireDateYyyymmdd = frame.header.tradeDateYyyymmdd;
            taxMatcher.openLot(lot);
        }

        benchmark.trackBatch(frame);

        st = runMarginPass(frame);
        ctx.marginOk = st == Status.OK;
        if ((frame.header.flags & WireTypes.BATCH_FLAG_MARGIN_CHECK) != 0 && !ctx.marginOk) return st;

        st = runHaircutPass(frame);
        if (st != Status.OK) return st;

        st = runCorporateActionPass(frame);
        if (st != Status.OK) return st;

        AllocationBatchSummary summary = new AllocationBatchSummary();
        summary.batchId = frame.header.deskId;
        summary.recordCount = ctx.recordCount;
        summary.totalQtyMilli = ctx.totalQtyMilli;
        summary.deskId = frame.header.deskId;
        summary.flags = frame.header.flags;
        summary.marginCleared = ctx.marginOk;
        summary.complianceCleared = ctx.complianceOk;
        AuditSpool.global().appendBatchSummary(summary);

        return Status.OK;
    }

    public Status runCompliancePass(BatchWireFrame frame) {
        RestrictionFilter.RestrictionFilterResult restricted = restrictions.filter(frame);
        if (restricted.status == Status.COMPLIANCE_REJECT) return Status.COMPLIANCE_REJECT;
        ComplianceRuleEngine.ComplianceResult cr = compliance.evaluateBatch(frame);
        return cr.passed ? Status.OK : Status.COMPLIANCE_REJECT;
    }

    public Status runMarginPass(BatchWireFrame frame) {
        MarginAggregator.MarginAggregatorResult res = margin.aggregate(frame);
        if (res.status != Status.OK && res.status != Status.MARGIN_BREACH) return res.status;
        if (res.breachCount > 0) return Status.MARGIN_BREACH;
        return Status.OK;
    }

    public Status runHaircutPass(BatchWireFrame frame) {
        HaircutCalculator.HaircutCalculatorSummary summary = haircut.computeFromBatch(frame);
        return summary.status;
    }

    public Status runCorporateActionPass(BatchWireFrame frame) {
        CorporateActionAdjuster.AdjustResult ar = corpAction.adjustBatch(frame);
        return ar.status;
    }

    public ProRataAllocator proRataAllocator() { return proRata; }
    public PositionLedger positionLedger() { return positions; }
    public ComplianceRuleEngine complianceEngine() { return compliance; }
    public MarginAggregator marginAggregator() { return margin; }
    public SettlementCalendar settlementCalendar() { return calendar; }
    public CounterpartyLimit counterpartyLimit() { return counterpartyLimit; }
    public TaxLotMatcher taxLotMatcher() { return taxMatcher; }
    public BenchmarkTracker benchmarkTracker() { return benchmark; }

    public Status runSettlementOnly(BatchWireFrame frame) {
        return calendar.computeBatch(frame).status;
    }

    public Status runRestrictionOnly(BatchWireFrame frame) {
        return restrictions.filter(frame).status;
    }

    public Status runFeeAccrualOnly(BatchWireFrame frame) {
        return fees.accrueBatch(frame).status;
    }

    public Status runLotSplitOnly(BatchWireFrame frame, LotSplitter.LotSplitPolicy policy) {
        return lotSplitter.splitBatch(frame, policy).status;
    }

    public DeskRunContext buildContext(BatchWireFrame frame) {
        DeskRunContext ctx = new DeskRunContext();
        runBatchDesks(frame, ctx);
        return ctx;
    }

    public boolean isFullyCleared(BatchWireFrame frame) {
        DeskRunContext ctx = new DeskRunContext();
        return runBatchDesks(frame, ctx) == Status.OK && ctx.marginOk && ctx.complianceOk;
    }

    public String summarizeRun(DeskRunContext ctx) {
        return "batch=" + ctx.batchId + " records=" + ctx.recordCount
                + " margin=" + ctx.marginOk + " compliance=" + ctx.complianceOk;
    }
}

File: legacy-cpp/src/desk/desk_registry.cc
#include "tkr/desk/desk_registry.h"

#include "tkr/desk/allocation_solver.h"
#include "tkr/desk/benchmark_tracker.h"
#include "tkr/desk/exposure_aggregator.h"
#include "tkr/desk/nav_calculator.h"
#include "tkr/desk/allocation_reconciler.h"
#include "tkr/desk/order_blotter.h"
#include "tkr/desk/portfolio_snapshot.h"
#include "tkr/desk/compliance_rule_engine.h"
#include "tkr/desk/corporate_action_adjuster.h"
#include "tkr/desk/counterparty_limit.h"
#include "tkr/desk/fee_accrual_engine.h"
#include "tkr/desk/haircut_calculator.h"
#include "tkr/desk/lot_splitter.h"
#include "tkr/desk/margin_aggregator.h"
#include "tkr/desk/position_ledger.h"
#include "tkr/desk/pro_rata_allocator.h"
#include "tkr/desk/restriction_filter.h"
#include "tkr/desk/settlement_calendar.h"
#include "tkr/desk/tax_lot_matcher.h"
#include "tkr/ledger/audit_spool.h"

namespace tkr {
namespace desk {

Status DeskRegistry::RunCompliancePass(const BatchWireFrame& frame) {
  RestrictionFilter filter(RestrictionFilterConfig{frame.header.desk_id, true});
  RestrictionFilterResult restricted = filter.Filter(frame);
  if (restricted.status == Status::kComplianceReject) {
    return Status::kComplianceReject;
  }

  ComplianceRuleEngine engine(
      ComplianceRuleEngineConfig{frame.header.desk_id, true});
  ComplianceRuleEngineResult res = engine.Evaluate(frame);
  if (res.status != Status::kOk) {
    return res.status;
  }
  if (res.rejected_records > 0) {
    return Status::kComplianceReject;
  }
  return Status::kOk;
}

Status DeskRegistry::RunMarginPass(const BatchWireFrame& frame) {
  MarginAggregator agg(MarginAggregatorConfig{frame.header.desk_id, 1000});
  MarginAggregatorResult res = agg.Aggregate(frame);
  if (res.status != Status::kOk && res.status != Status::kMarginBreach) {
    return res.status;
  }
  if (res.breach_count > 0) {
    return Status::kMarginBreach;
  }
  return Status::kOk;
}

Status DeskRegistry::RunBatchDesks(BatchWireFrame& frame, DeskRunContext* ctx) {
  if (ctx == nullptr) {
    return Status::kBoundsError;
  }

  ctx->batch_id = frame.header.desk_id;
  ctx->record_count = static_cast<std::uint32_t>(frame.records.size());
  ctx->total_qty_milli = 0;
  for (const WireBatchRecord& rec : frame.records) {
    ctx->total_qty_milli += rec.qty_milli;
  }

  Status compliance = RunCompliancePass(frame);
  ctx->compliance_ok = compliance == Status::kOk;
  if ((frame.header.flags & kBatchFlagComplianceHold) != 0 &&
      !ctx->compliance_ok) {
    return compliance;
  }

  CounterpartyLimit cp_limit(CounterpartyLimitConfig{frame.header.desk_id, true});
  CounterpartyLimitResult cp_result = cp_limit.Evaluate(frame);
  if (cp_result.status == Status::kComplianceReject) {
    ctx->compliance_ok = false;
    return Status::kComplianceReject;
  }

  SettlementCalendar calendar(SettlementCalendarConfig{
      MarketRegion::kUsEquity, SettlementCycle::kT2, true});
  SettlementCalendarResult settle_result = calendar.ComputeBatch(frame);
  if (settle_result.status != Status::kOk) {
    return settle_result.status;
  }

  LotSplitter splitter(LotSplitterConfig{1000, true});
  LotSplitterResult split = splitter.SplitBatch(frame, LotSplitPolicy::kProRataByWeight);
  if (split.status != Status::kOk) {
    return split.status;
  }

  ProRataAllocator allocator(ProRataAllocatorConfig{true});
  ProRataAllocatorResult alloc = allocator.Allocate(frame);
  if (alloc.status != Status::kOk) {
    return alloc.status;
  }

  AllocationSolver solver(AllocationSolverConfig{true, 1000000});
  AllocationSolverResult solved = solver.Solve(frame, alloc.slices);
  if (solved.status != Status::kOk) {
    return solved.status;
  }

  PositionLedger ledger(PositionLedgerConfig{frame.header.desk_id});
  Status ledger_st = ledger.ApplyBatch(frame, solved.final_slices);
  if (ledger_st != Status::kOk) {
    return ledger_st;
  }

  FeeAccrualEngine fees(FeeAccrualEngineConfig{frame.header.desk_id, frame.header.trade_date_yyyymmdd});
  FeeAccrualEngineResult fee_res = fees.AccrueBatch(frame);
  if (fee_res.status != Status::kOk) {
    return fee_res.status;
  }

  TaxLotMatcher matcher(TaxLotMatcherConfig{TaxLotMatchMethod::kFifo});
  for (const WireBatchRecord& rec : frame.records) {
    TaxLot lot{};
    lot.lot_id = rec.record_id;
    lot.account_id = rec.account_id;
    lot.symbol_id = rec.symbol_id;
    lot.qty_milli = static_cast<std::int64_t>(rec.qty_milli);
    lot.cost_basis_cents =
        (static_cast<std::int64_t>(rec.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;
    lot.acquire_date_yyyymmdd = frame.header.trade_date_yyyymmdd;
    lot.closed = false;
    matcher.OpenLot(lot);
  }

  BenchmarkTracker tracker(BenchmarkTrackerConfig{1, 100});
  BenchmarkTrackerResult bench = tracker.TrackBatch(frame);

  NavCalculator nav(25, 150);
  NavSnapshot nav_snap = nav.Compute(frame);
  ExposureAggregator exposure(frame.header.desk_id, true);
  ExposureReport exposure_report = exposure.Aggregate(frame);
  PortfolioSnapshotBuilder portfolio(frame.header.desk_id, 100000000);
  PortfolioSnapshot portfolio_snap = portfolio.Build(frame);
  OrderBlotter blotter(frame.header.trade_date_yyyymmdd);
  blotter.IngestBatch(frame);
  AllocationReconciler reconciler;
  ReconcileReport reconcile_report = reconciler.ReconcileBatch(frame, alloc.slices);

  Status margin = RunMarginPass(frame);
  ctx->margin_ok = margin == Status::kOk;
  if ((frame.header.flags & kBatchFlagMarginCheck) != 0 && !ctx->margin_ok) {
    return margin;
  }

  AllocationBatchSummary summary{};
  summary.batch_id = frame.header.desk_id;
  summary.record_count = ctx->record_count;
  summary.total_qty_milli = ctx->total_qty_milli;
  summary.desk_id = frame.header.desk_id;
  summary.flags = frame.header.flags;
  summary.margin_cleared = ctx->margin_ok;
  summary.compliance_cleared = ctx->compliance_ok;
  ledger::GlobalAuditSpool().AppendBatchSummary(summary);

  (void)bench;
  (void)split;
  (void)settle_result;
  (void)nav_snap;
  (void)exposure_report;
  (void)portfolio_snap;
  (void)blotter;
  (void)reconcile_report;
  return Status::kOk;
}

Status DeskRegistry::RunHaircutPass(const BatchWireFrame& frame) {
  HaircutCalculator calc(HaircutCalculatorConfig{frame.header.desk_id, true});
  std::vector<HaircutInput> inputs;
  inputs.reserve(frame.records.size());

  std::int64_t total_gross = 0;
  for (const WireBatchRecord& rec : frame.records) {
    total_gross +=
        (static_cast<std::int64_t>(rec.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;
  }

  for (const WireBatchRecord& rec : frame.records) {
    HaircutInput input{};
    input.symbol_id = rec.symbol_id;
    input.asset_class = AssetClass::kEquity;
    input.rating = CreditRating::kA;
    input.market_value_cents =
        (static_cast<std::int64_t>(rec.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;
    if (total_gross > 0) {
      input.portfolio_weight_bp = static_cast<std::uint32_t>(
          (input.market_value_cents * 10000) / total_gross);
    }
    inputs.push_back(input);
  }

  HaircutCalculatorSummary summary = calc.Compute(inputs);
  return summary.status;
}

Status DeskRegistry::RunCorporateActionPass(const BatchWireFrame& frame) {
  CorporateActionAdjuster adjuster(
      CorporateActionAdjusterConfig{frame.header.desk_id});
  CorporateActionAdjusterResult result = adjuster.ApplyToBatch(frame);
  return result.status;
}

}  // namespace desk
}  // namespace tkr
